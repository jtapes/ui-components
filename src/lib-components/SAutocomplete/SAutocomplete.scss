@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-autocomplete {
  position: relative;
  border-bottom: 1px solid $color--border;
  padding-top: 3px;
  background-color: #fff;
  transition-property: border;
  transition-duration: $animate--time;

  .s-content-loader svg {
    width: 100%;
    height: 140px;
  }

  .s-input {
    border-color: transparent;
    background-color: transparent;

    &__append {
      margin-right: 0;
    }

    &:focus-within {
      box-shadow: none;
    }

    &:hover {
      border-color: transparent;
    }
  }

  &:hover {
    &:not(.s-autocomplete--readonly) {
      border-color: $color--primary-accent;
    }
  }

  &.s-autocomplete--rounded {
    border: 1px solid $color--border;
    border-radius: $size--24;
    padding: 0 $size--4 0 0;
  }

  &.s-autocomplete--open {
    border-color: $color--border;
    border-radius: $size--24 $size--24 0 0;
    box-shadow: 0 0 0 3px $color--outline;

    .s-input {
      border-radius: $size--24 $size--24 0 0;
    }

    &:not(.s-autocomplete--rounded) {
      margin-right: -$size--16;
      margin-left: -$size--16;
      padding-right: $size--16;
      padding-left: $size--16;
      width: auto;
    }
  }

  &.s-autocomplete--sm {
    .s-autocomplete-suggestions {
      &__item {
        padding: $size--8 $size--20;
      }

      &__empty {
        padding: $size--8 $size--20;
      }

      &__list {
        max-height: 140px;
      }
    }
  }

  &.s-autocomplete--readonly {
    &.s-autocomplete--rounded {
      border-color: $color--background;
      background-color: $color--background;
    }
  }

  &.s-autocomplete--error {
    border-color: $color--danger;

    &.s-autocomplete--open {
      border-color: transparent;
    }
  }
}

.s-autocomplete-suggestions {
  @include z-index(modal);

  position: absolute;
  left: 0;
  right: 0;
  top: 100%;
  border-top: 1px solid $color--border;
  border-radius: 0 0 $size--24 $size--24;
  padding: $size--20 $size--16;
  box-shadow: 0 3px 0 $color--outline, 3px 3px 0 $color--outline, -3px 3px 0 $color--outline;
  background-color: $color--white;

  &__list {
    @include scrollbar;

    overflow-y: auto;
    margin-right: -$size--12;
    margin-left: -$size--16;
    max-height: 160px;
  }

  &__item {
    padding: $size--12 $size--16;
    cursor: pointer;

    &:hover {
      background-color: $color--complimentary-bg-light;
    }
  }

  &__empty {
    padding: $size--12 $size--16;
  }
}
