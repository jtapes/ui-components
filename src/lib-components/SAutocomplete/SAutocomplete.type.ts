export type SAutocompleteItem<T> = {
  text: string
  value: T
}

export type SAutocompleteItemValue = { [key: string]: any } | string | number
