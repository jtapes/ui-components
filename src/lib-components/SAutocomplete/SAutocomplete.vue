<template>
  <div
    ref="wrapper"
    class="s-autocomplete"
    :class="{
      's-autocomplete--open' : opened,
      's-autocomplete--sm': size === 'sm',
      's-autocomplete--rounded': rounded,
      's-autocomplete--readonly': state === 'readonly',
      's-autocomplete--error': state === 'error'
    }"
  >
    <s-input
      :value="value"
      :size="size"
      v-bind="$attrs"
      :clearable="clearable"
      :rounded="rounded"
      :state="state"
      v-on="listeners"
      @focus="handleFocus"
    >
      <template v-if="!iconHidden" #append>
        <!-- @slot Слот для кастомной иконки -->
        <slot v-if="!loading" name="icon">
          <svg-icon name="search-24-32" class="icon-32" />
        </slot>
        <svg
          v-else
          class="s-loader-spinner icon-32 color-font-secondary"
          viewBox="0 0 32 32"
        >
          <circle
            class="s-loader-spinner__path"
            cx="16"
            cy="16"
            r="10"
            fill="none"
            stroke="currentColor"
          />
        </svg>
      </template>
    </s-input>
    <transition name="s-zoom-in-top">
      <div
        v-show="opened"
        ref="dropdown"
        class="s-autocomplete-suggestions"
        :style="{ ...appendedToBodyPosition, ...appendedToBodySize }"
      >
        <div
          v-if="!loading"
          class="s-autocomplete-suggestions__list"
        >
          <div
            v-for="(item, index) in mappedItems"
            :key="index"
            class="s-autocomplete-suggestions__item"
            @click="handleSelectedChange(item)"
          >
            <!-- @slot Слот для элемента списка -->
            <slot :item="item">
              <div class="text-sm">
                {{ item.text }}
              </div>
            </slot>
          </div>
          <div
            v-if="!items.length"
            class="s-autocomplete-suggestions__empty text-sm color-font-secondary"
          >
            Ничего не найдено
          </div>
        </div>
        <div v-else class="s-autocomplete-suggestions__loading">
          <s-content-loader height="140" width="100" preserve-aspect-ratio="none">
            <rect
              x="0"
              y="0"
              width="100%"
              height="20"
            />
            <rect
              x="0"
              y="40"
              width="100%"
              height="20"
            />
            <rect
              x="0"
              y="80"
              width="100%"
              height="20"
            />
            <rect
              x="0"
              y="120"
              width="100%"
              height="20"
            />
          </s-content-loader>
        </div>
      </div>
    </transition>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop, Model } from 'vue-property-decorator'
import SInput from '@/lib-components/SInput/SInput.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import SContentLoader from '@/lib-components/SContentLoader/SContentLoader.vue'
import {
  SAutocompleteItem,
  SAutocompleteItemValue
} from '@/lib-components/SAutocomplete/SAutocomplete.type'

type AppendedToBodySize = {
  width?: string
}

type AppendedToBodyPosition = {
  top?: string
  left?: string
}

/** Компонент для поискового инпута с подгрузкой списка */
@Component({
  components: {
    SInput,
    SvgIcon,
    SContentLoader
  },
})
export default class SAutocomplete extends Vue {
  @Model('change')
  value!: SAutocompleteItemValue

  appendedToBodyPosition: AppendedToBodyPosition = {}
  appendedToBodySize: AppendedToBodySize = {}

  /** Массив данных для выпадающего списка */
  @Prop({ type: Array, default: () => [] })
  items!: SAutocompleteItem<SAutocompleteItemValue>[] | number[] | string[]

  /** Идентификатор предзагрузки */
  @Prop(Boolean)
  loading?: boolean

  /** Размер инпута */
  @Prop(String)
  size?: string

  /** Свойство округления инпута */
  @Prop({ type: Boolean, default: true })
  rounded!: boolean

  /** Свойство очистки инпута */
  @Prop(Boolean)
  clearable?: boolean

  /** Название свойства для лэйбла */
  @Prop({ type: String, default: 'text' })
  itemText!: string

  /** Название свойства для значения */
  @Prop({ type: String, default: 'value' })
  itemValue!: string

  /** Свойство выноса списка в body документа */
  @Prop({ type: Boolean, default: false })
  appendToBody!: boolean

  /** Свойство для скрытия иконки */
  @Prop({ type: Boolean, default: false })
  iconHidden!: boolean

  @Prop(String)
  state!: string

  opened = false

  get listeners() {
    return {
      ...this.$listeners,
      change: this.handleInput,
    }
  }

  get mappedItems(): SAutocompleteItem<SAutocompleteItemValue>[] {
    return this.items.map((item: SAutocompleteItemValue) => {
      const text = typeof item === 'object' ? item[this.itemText] : String(item)
      const value = typeof item === 'object' ? item[this.itemValue] : item
      return {
        text,
        value,
      }
    })
  }

  handleInput(value: string): void {
    // TODO: Более изощренного решения для избавления от двойного эмита без переписывания функцианала s-input пока нет

    if (value !== this.value) {
      this.opened = true
      if (this.appendToBody) this.prepareDropdown()
      this.$emit('change', value)
    }
  }

  handleFocus(): void {
    if (this.appendToBody) this.prepareDropdown()
    if (this.value) {
      this.opened = true
    }
  }

  handleSelectedChange(value: SAutocompleteItem<SAutocompleteItemValue>): void {
    this.opened = false

    this.$emit('select-item', value)
    this.$emit('change', value.text)
  }

  handleOpenedChange(): void {
    this.opened = !this.opened
    if (this.opened && this.appendToBody) {
      this.prepareDropdown()
    }
  }

  prepareDropdown(): void {
    if (!this.$isServer) {
      this.$nextTick(async () => {
        if (this.$refs.wrapper && this.$refs.dropdown) {
          const slotRect = (this.$refs.wrapper as Element).getBoundingClientRect()
          this.appendedToBodySize = {
            width: `${slotRect.width}px`,
          }
          this.appendedToBodyPosition = {
            top: `${slotRect.top + slotRect.height + window.scrollY}px`,
            left: `${slotRect.left}px`,
          }
        }
      })
    }
  }

  close(e: MouseEvent): void {
    const element = e.target as Node

    if (!this.$el.contains(element)) {
      this.opened = false
    }
  }

  mounted(): void {
    document.addEventListener('click', this.close, true)
    if (this.appendToBody) {
      document.body.append((this.$refs.dropdown as Node))
      this.prepareDropdown()
    }
  }

  beforeDestroy(): void {
    document.removeEventListener('click', this.close, true)
  }
}
</script>
