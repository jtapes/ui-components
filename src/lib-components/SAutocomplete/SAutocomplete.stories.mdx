import { action } from '@storybook/addon-actions'
import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Props,
  Canvas,
} from '@storybook/addon-docs'
import SAutocomplete from '@/lib-components/SAutocomplete/SAutocomplete.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import SButton from '@/lib-components/SButton/SButton.vue'
import { SvgMdxArgType } from '@/storybook/components/icon/SvgIconArgType'

import results from '/.jest-test-results.json';

<Meta
  title="DESIGN SYSTEM/Atom/Autocomplete"
  component={SAutocomplete}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    items: {
      control: 'array',
      defaultValue: [
        { text: 'Значение 1', value: 1 },
        { text: 'Значение 2', value: 2 },
        { text: 'Значение 3', value: 3 },
        { text: 'Значение 4', value: 4 },
        { text: 'Значение 5', value: 5 },
        { text: 'Значение 6', value: 6 },
        { text: 'Значение 7', value: 7 },
        { text: 'Значение 8', value: 8 }
      ]
    },
    size: {
      options: [null, 'sm'],
      control: { type: 'select' },
    },
    rounded: { control: 'boolean', defaultValue: true },
    loading: { control: 'boolean', defaultValue: false },
    icon: SvgMdxArgType,
    appendToBody: { control: 'boolean', defaultValue: false },
    clearable: { control: 'boolean', defaultValue: false },
    iconHidden: { control: 'boolean', defaultValue: false },
    state: {
      control: { type: 'select' },
      options: [
        'readonly',
        'error',
      ]
    },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/IIeFkxNoHGcaqNHlVeNrhW/MVP-3-Market-UI-Kit?node-id=846%3A1100',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# SAutocomplete

<Description of={SAutocomplete} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SAutocomplete, SvgIcon, SButton },
  template:
    `<div style="padding: 0 16px; background-color: #eee">
      <SAutocomplete :items="items" :size="size" :rounded="rounded" :loading="loading" placeholder="Введите"
:appendToBody="appendToBody" :clearable="clearable" :iconHidden="iconHidden" v-model="txt" :state="state" placeholder="Placeholder">
<template #icon><s-button mode="search" size="sm" class="ml-8"circle><svg-icon name="search-24-32" class="icon-32"></svg-icon></s-button></template>
    </SAutocomplete>
</div>`,
  data() {
  return {
    txt: ''
  }}
})

<Canvas withToolbar withSource="none">
  <Story name="SAutocomplete" args={{  }}>
    {Template.bind({})}
  </Story>
</Canvas>

## Дефолтный пример

<Source
  language="html"
  code={`
    <SAutocomplete :items="items" :size="size" :rounded="rounded" :loading="loading" placeholder="Поиск"
:appendToBody="appendToBody" :clearable="clearable" :iconHidden="iconHidden">
      <template #icon v-if="icon"><svg-icon :name="icon" class="icon-32" /></template>
</SAutocomplete>
  `}
/>

<ArgsTable of={SAutocomplete} />
