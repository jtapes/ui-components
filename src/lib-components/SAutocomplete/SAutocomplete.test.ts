import { mount } from '@vue/test-utils'
import SAutocomplete from '@/lib-components/SAutocomplete/SAutocomplete.vue'

describe('SAutocomplete', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SAutocomplete)
    expect(wrapper.element).toMatchSnapshot()
  })

  test('$emit() работает', async () => {
    const wrapper = mount(SAutocomplete)
    const input = wrapper.find('input')
    await input.setValue('test text')
    expect(wrapper.emitted('change')).toBeTruthy()
  })
})
