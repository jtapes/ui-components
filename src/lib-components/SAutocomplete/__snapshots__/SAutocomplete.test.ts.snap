// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SAutocomplete Is Vue snapshot 1`] = `
<div
  class="s-autocomplete s-autocomplete--rounded"
>
  <div
    class="s-input s-input--text s-input--rounded"
  >
    <!---->
     
    <input
      class="s-input__inner s-input__inner--appended"
      type="text"
    />
     
    <!---->
     
    <!---->
     
    <div
      class="s-input__append"
    >
      <svg
        class="svg-icon icon-32 svg-icon--search-24-32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          href="#search-24-32"
          xlink:href="#search-24-32"
        />
      </svg>
    </div>
  </div>
   
  <transition-stub
    name="s-zoom-in-top"
  >
    <div
      class="s-autocomplete-suggestions"
      style="display: none;"
    >
      <div
        class="s-autocomplete-suggestions__list"
      >
         
        <div
          class="s-autocomplete-suggestions__empty text-sm color-font-secondary"
        >
          
          Ничего не найдено
        
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;
